File: solutions/397.integer-replacement/integer-replacement.java
class Solution {
    public int integerReplacement(int n) {
        if(n < 4) return n - 1;
        long N = n;
        int res = 0;
        while(N != 1) {
            if(N % 2 == 0) {
                N >>= 1;
            } else {
                if(N == 3) return res + 2;
                N = (N & 2) == 2 ? N + 1 : N - 1;
            }
            res++;
        }
        return res;
    }
}

File: solutions/397.integer-replacement/integer-replacement.cpp
class Solution {
public:
    int integerReplacement(int n) {
        return (int)longReplacement(n);
    }
private: 
    long longReplacement(long n) {
        if(n < 3) return n - 1;
        if(n % 2 == 0) return longReplacement(n / 2) + 1;
        return min(longReplacement(n - 1), longReplacement(n + 1)) + 1;
    }
};